gotocompany/shield/v1beta1/servicedata.proto

```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+
+package gotocompany.shield.v1beta1;
+
+import "google/api/annotations.proto";
+import "protoc-gen-openapiv2/options/annotations.proto";
+import "validate/validate.proto";
+
+option go_package = "github.com/goto/proton/shield/v1;shieldv1beta1";
+option java_outer_classname = "ServiceData";
+option java_package = "com.gotocompany.proton.shield.v1beta1";
+// These annotations are used when generating the OpenAPI file.
+option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
+  info: {
+    title: "Service Data",
+    version: "0.1.0";
+  };
+  schemes: HTTP;
+};
+
+service ServiceDataService {
+   // Service Data
+   rpc CreateServiceDataKey(CreateServiceDataKeyRequest) returns (CreateServiceDataKeyResponse) {
+    option (google.api.http) = {
+      post: "/v1beta1/servicedata",
+      body: "body"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      tags: "Service Data";
+      summary: "Create Service Data Key";
+    };
+  }
+
+  rpc UpsertUserServiceData(UpsertUserServiceDataRequest) returns (UpsertUserServiceDataResponse) {
+    option (google.api.http) = {
+      put: "/v1beta1/users/{id}/servicedata",
+      body: "body"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      tags: "Service Data";
+      summary: "Upsert User Service Data";
+    };
+  }
+    
+  rpc UpsertGroupServiceData(UpsertGroupServiceDataRequest) returns (UpsertGroupServiceDataResponse) {
+    option (google.api.http) = {
+      put: "/v1beta1/groups/{id}/servicedata",
+      body: "body"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      tags: "Service Data";
+      summary: "Upsert Group Service Data";
+    };
+  }
+}
+
+message ServiceDataKeyRequestBody {
+  string project = 1;
+  string key = 2;
+  string description = 3;
+}
+
+message ServiceDataKey {
+  string urn = 1;
+  string id = 2;
+}
+
+message CreateServiceDataKeyRequest {
+  ServiceDataKeyRequestBody body = 1;
+}
+
+message CreateServiceDataKeyResponse {
+  ServiceDataKey service_data_key = 1;
+}
+
+message UpsertServiceDataRequestBody {
+  string project = 1;
+  map<string, string> data = 2 [(validate.rules).map = {min_pairs: 1, max_pairs: 1}];
+}
+  
+message UpsertUserServiceDataRequest {
+  string id = 1;
```

Comment:
For upsert, we can remove this id, lets not use the `id` but just rely on the key in the data map
the uniqueness should be the combination of (project, entity id, key)
```suggestion
```

Reply:
so are we sending entity id in the body?

Reply:
So this is entity id?
I thought it is the service data id.

if it is entity id, can you rename it to entity_id or user_id ?

Reply:
It is the id that being sent in param. Sure, will rename it